### PR TITLE
ODC-6778: Bump openshift/api to get devcatalog disable types cuztomization changes

### PR DESCRIPTION
**Pull request by debsmita1**
**Story:**
https://issues.redhat.com/browse/ODC-6778

Add dev catalog types customization API based on https://github.com/openshift/enhancements/pull/1206

**Test setup:**
This PR bumps the openshift/api, includes changes from https://github.com/openshift/api/pull/1256, and finally copies the data from the cluster Console configuration to the console-config ConfigMap.

To test this change launch a cluster with launch openshift/console-operator#676

Open https://cluster/k8s/cluster/operator.openshift.io~v1~Console/cluster/yaml and configure customization like:
```
  customization:
    developerCatalog:
      types:
         state: Disabled
         disabled:
         - HelmChart
         - Devfile
```
Open https://cluster/k8s/ns/openshift-console/configmaps/console-config and check data key console-config.yaml. The YAML should contain the disableTypes after a few seconds.
```
customization:
    developerCatalog:
      types:
         state: Disabled
         disabled:
         - HelmChart
         - Devfile

### CODE REVIEW

**Comment by Kartikey-star** [comment]
@debsmita1 I am bumping api  as per https://github.com/openshift/api/pull/1213 PR approval in . Does  the changes you would like to see in  vendor for openshift/api present before this PR approval.
If that's the case then i can bump the api version in https://github.com/openshift/console-operator/pull/674 and you can add the other changes in this PR.

**Review by jhadvig — COMMENTED** [review]
@debsmita1 thanks for the PR, but I dont see the api bump in any commit :-/

File: pkg/console/subresource/consoleserver/types.go
Comment (jhadvig) [inline pkg/console/subresource/consoleserver/types.go:115]: why we need list of disabled types ? I would assume that if the type is not enabled, then its disabled?

**Review by jhadvig — CHANGES_REQUESTED** [review]
Adding WIP prefix to the PR, since API not in place.

File: go.mod
Comment (debsmita1) [reply go.mod:28]: I replaced this so that we can test the console-operator changes and the console changes together on a dev cluster.
I will remove this once the api gets merged

File: pkg/console/subresource/consoleserver/types.go
Comment (jhadvig) [reply pkg/console/subresource/consoleserver/types.go:103]: eg.
```go
// DeveloperConsoleCatalogTypes contains list of enabled and disabled catalog types.
// If any sub-catalog types are mentioned in both `enabled` and `disabled` list, the sub-catalog type(s) should be hidden. ...
```

File: pkg/console/subresource/consoleserver/types.go
Comment (debsmita1) [reply pkg/console/subresource/consoleserver/types.go:116]: We considered both because, for ex, if there are 10 sub-catalog types and the admin wants to enable 8 of them, then instead of listing all the types in `enabled` the admin can simply list the remaining 2 in the disabled list.

**Comment by jerolimov** [comment]
> /hold lets wait till enhancement and api PRs are merged, so we dont launch CI clusters unnecessarily.

Sorry, currently I/we don't understand why the e2e are failing. To get this in asap it is required to understand if the API changes are working as part of this operator. So until the builds are green again, it's really helpful (required) to run them.

/unhold

**Comment by debsmita1** [comment]
/hold
till the api PR https://github.com/openshift/api/pull/1256 gets in

File: pkg/console/subresource/consoleserver/config_builder.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder.go:388]: Looks same to me, categories is an empty error, or you create one. Isn't this enough here?

```suggestion
	conf.DeveloperCatalog = &DeveloperConsoleCatalogCustomization{
		Categories: &categories,
		Types:      catalogTypes,
	}
```

File: pkg/console/subresource/consoleserver/config_builder.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder.go:361]: The "enum" is just a typed string, comparing this with the local CatalogTypeDisabled needs a cast, but it might be easier to use the const from operatorv1:

```suggestion
			if b.devCatalogCustomization.Types.State == operatorv1.CatalogTypeDisabled && b.devCatalogCustomization.Types.Disabled != nil {
				disabledTypes = *b.devCatalogCustomization.Types.Disabled
			}

			if b.devCatalogCustomization.Types.State == operatorv1.CatalogTypeEnabled && b.devCatalogCustomization.Types.Enabled != nil {
				enabledTypes = *b.devCatalogCustomization.Types.Enabled
			}
```

There are some more spots for CatalogTypeDisabled and CatalogTypeEnabled

Comment (jerolimov) [reply pkg/console/subresource/consoleserver/config_builder.go:361]: But one more note here. Isn't it enough to copy the complete `catalogTypes` like this?

```go
		if !reflect.DeepEqual(b.devCatalogCustomization.Types, operatorv1.DeveloperConsoleCatalogTypesState{}) {
			catalogTypes = DeveloperConsoleCatalogTypesState{
				State:    CatalogTypesState(b.devCatalogCustomization.Types.State),
				Enabled:  b.devCatalogCustomization.Types.Enabled,
				Disabled: b.devCatalogCustomization.Types.Disabled,
			}
		}
```

**Review by jerolimov — CHANGES_REQUESTED** [review]

File: pkg/console/subresource/consoleserver/types.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/types.go:123]: Hmm, missed that in the API change, but this defines the "Catalog Types", and contains a state, so I believe we should drop the state suffix from this type definition, or?

```suggestion
// DeveloperConsoleCatalogTypes defines the state of the sub-catalog types.
type DeveloperConsoleCatalogTypes struct {
```

Comment (debsmita1) [reply pkg/console/subresource/consoleserver/types.go:123]: here I am trying to define the state of the catalog types

File: pkg/console/subresource/consoleserver/config_builder.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder.go:350]: One check should be enough 😏

**Review by jerolimov — CHANGES_REQUESTED** [review]

**Comment by jerolimov** [comment]
> /hold till the api PR [openshift/api#1256](https://github.com/openshift/api/pull/1256) gets in

it is merged :tada: and go.mod refers just the latest version

/unhold

File: pkg/console/subresource/consoleserver/config_builder.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder.go:323]: Can you define `categories []DeveloperConsoleCatalogCategory` here and create the list via make in the `if b.devCatalogCustomization.Categories != nil` below?

File: pkg/console/subresource/consoleserver/types.go
Comment (jerolimov) [reply pkg/console/subresource/consoleserver/types.go:123]: I was thinking that `State CatalogTypesState `yaml:"state,omitempty"`` is "the state" and this is "more", but its fine, we can keep this.

File: pkg/console/subresource/consoleserver/config_builder_test.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder_test.go:282]:
```suggestion
					Types: v1.DeveloperConsoleCatalogTypesState{State: v1.CatalogTypeDisabled, Disabled: &[]string{"type1", "type2"}},
```

File: pkg/console/subresource/consoleserver/config_builder_test.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder_test.go:316]:
```suggestion
					Types: v1.DeveloperConsoleCatalogTypesState{State: v1.CatalogTypeEnabled, Enabled: &[]string{}},
```

File: pkg/console/subresource/consoleserver/config_builder_test.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder_test.go:419]:
```suggestion
							State: v1.PerspectiveAccessReview,
```

File: pkg/console/subresource/consoleserver/config_builder_test.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder_test.go:428]:
```suggestion
							State: v1.PerspectiveDisabled,
```

File: pkg/console/subresource/consoleserver/config_builder_test.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder_test.go:785]:
```suggestion
					Types: v1.DeveloperConsoleCatalogTypesState{State: v1.CatalogTypeDisabled, Disabled: &[]string{"type1", "type2"}},
```

File: pkg/console/subresource/consoleserver/config_builder_test.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder_test.go:816]:
```suggestion
					Types: v1.DeveloperConsoleCatalogTypesState{State: operatorv1.CatalogTypeEnabled, Enabled: &[]string{}},
```

File: pkg/console/subresource/consoleserver/config_builder_test.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder_test.go:955]:
```suggestion
							State: v1.PerspectiveAccessReview,
```

File: pkg/console/subresource/consoleserver/config_builder_test.go
Comment (jerolimov) [inline pkg/console/subresource/consoleserver/config_builder_test.go:964]:
```suggestion
							State: v1.PerspectiveDisabled,
```

**Review by jerolimov — COMMENTED** [review]
These consts doesn't need a cast:

**Review by jerolimov — APPROVED** [review]
Tested this on a cluster and it works fine :+1: 

/lgtm
/label qe-approved

Propagate from the epic:

/label px-approved
/label docs-approved

**Comment by jhadvig** [comment]
/hold
All the e2e are failing due to:
```
Console.operator.openshift.io "cluster" is invalid: spec.customization.developerCatalog.types.state: Required value
```

@jerolimov also I dont thing that we should be adding `qe-approved` labels ourselves. The no-FF process has it's meaning and should be followed for console related repositories.

**Comment by jerolimov** [comment]
Hey @JoelSpeed @deads2k @jhadvig @sanketpathak,

@Debsmita and I have tested this change together, which includes https://github.com/openshift/api/pull/1310 now, on a cluster bot instance. Sorry I missed recording it. :man_shrugging: 

Manual testing was successful :+1: and also the e2e tests run fine now. :partying_face: 

Some details:

1. When defining just `types: {}` our UI says the state is missing (it's correct), but you can save it anyway (like the cli would do it) and the state was automatically set to `Enabled`.
2. We've verified that it's possible to save `Enabled`, `Disabled` catalog lists and that mixing `state: Disabled` and an `enabled` list, etc. doesn't work. All works fine. We've tested also the CEL rules on perspectives not on k8s 1.25 and it works as expected.
3. And finally, we checked that the console `ConfigMap` was updated and that the JS global variable (`window.SERVER_FLAGS.developerCatalogTypes`) in the UI was exported correctly.

We need to merge https://github.com/openshift/api/pull/1310 first and update this here one more time.

**Comment by jerolimov** [comment]
Tested again on a cluster bot instance and works fine:

https://user-images.githubusercontent.com/139310/193407434-0357a338-bc8b-4be2-ae2d-d5bcb598cfd5.mp4

/lgtm

**Comment by invincibleJai** [comment]
/hold cancel

as PR got +1 from @jhadvig

**Pull request merged**
